tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc

```diff
@@ -0,0 +1,674 @@
+          "status: ",
+          cudaGetErrorString(err));
+    }
+    {
+      Tensor temp_storage2;
+      TF_RETURN_IF_ERROR(context->allocate_temp(
+          DT_INT8, TensorShape({static_cast<int64>(temp_storage_bytes)}),
+          &temp_storage2));
+      err = gpuprim::DeviceScan::InclusiveSum(
+          /*d_temp_storage=*/temp_storage2.flat<int8>().data(),
+          /*temp_storage_bytes=*/temp_storage_bytes,
+          /*d_in=*/empty_row_indicator,
+          /*d_out=*/num_empty_rows_through.data(),
+          /*num_items=*/dense_rows,
+          /*stream=*/cu_stream);
+      if (err != 0) {
+        return errors::Internal(
+            "SparseFillEmptyRows: Could not launch "
+            "gpuprim::DeviceScan::ExclusiveSum to scan empty_row_indicator, "
+            "temp_storage_bytes:",
+            temp_storage_bytes, ", status: ", cudaGetErrorString(err));
+      }
+    }
+
+    ScratchSpace<Tindex> num_empty_rows_host(context, 1, /*on_host=*/true);
+    if (!stream
+             ->ThenMemcpy(num_empty_rows_host.mutable_data(),
+                          se::DeviceMemoryBase(
+                              num_empty_rows_through.data() + (dense_rows - 1),
+                              sizeof(*num_empty_rows_host.data())),
+                          sizeof(*num_empty_rows_host.data()))
+             .ok()) {
+      return errors::Internal("Failed to copy num_empty_rows to host");
+    }
+
+    ScratchSpace<int> rows_are_not_ordered_host(context, 1, /*on_host=*/true);
+    if (!stream
+             ->ThenMemcpy(rows_are_not_ordered_host.mutable_data(),
+                          rows_are_not_ordered_gpu_memory,
+                          sizeof(*rows_are_not_ordered_host.data()))
+             .ok()) {
+      return errors::Internal("Failed to copy rows_are_not_ordered to host");
+    }
+
+    ScratchSpace<int> first_invalid_index_host(context, 1, /*on_host=*/true);
+    if (!stream
+             ->ThenMemcpy(first_invalid_index_host.mutable_data(),
+                          first_invalid_index_gpu_memory,
+                          sizeof(*first_invalid_index_host.data()))
+             .ok()) {
+      return errors::Internal("Failed to copy first_invalid_index to host");
+    }
+
+    // We must wait for num_empty_rows and rows_are_not_ordered to be copied to
+    // the host, so we enqueue the remainder of the computation onto the stream
+    // asynchronously to avoid stalling execution.
+    auto async_finish_computation =
+        [this, context, kReverseIndexMapOutput, kAllIndicesValid, index_type, N,
+         rank, indices, values, default_value, dense_rows, num_empty_rows_host,
+         rows_are_not_ordered_host, first_invalid_index_host,
+         num_empty_rows_through_t, num_empty_rows_through, input_row_ends_t,
+         input_row_ends, empty_row_indicator_t, empty_row_indicator,
+         done]() -> void {
+      CHECK(done);  // Crash OK
```

Comment:
Can you please proactively change these to be `DCHECK`s to avoid accidentally holding up the merge?

Reply:
Done.